feat:셀프 조인을 통한 Qna 대댓글 구현

cyj1127 opened the pull request:
parent 유무로 댓글,대댓글 분기 로직 구현

Commit:
feat:셀프 조인을 통한 Qna 대댓글 구현

parent 유무로 댓글,대댓글 분기 로직 구현

ReadAlien commented on src/main/java/com/gagoo/thiscoding/domain/maria/reply/controller/ReplyWriteController.java:
여기서 필요없는 코드는 삭제해야할 거 같아요

cyj1127 replied:
확인 후 삭제했습니다.

ReadAlien commented on src/main/java/com/gagoo/thiscoding/domain/maria/reply/domain/Reply.java:
여기 도메인에서도 전 pr에 남긴 리뷰처럼 순수도메인을 지키는 방식을 찾아야할 거 같아요

cyj1127 replied:
필드에 ReplyEntity 타입을 Reply타입으로 바꿔서 순수도메인을 유지할 수 있게 코드 수정하였습니다.

ReadAlien commented on src/main/java/com/gagoo/thiscoding/domain/maria/reply/domain/dto/ReplyCreate.java:
parent보다는 parentId가 더 적합한 이름인 거 같은데 바뀐 이유가 있나요?

cyj1127 replied:
셀프조인 후 결과값을 parent_id필드로 db에 저장하지만 parent객체와 조인을 하는 점에서 parentId보단 parent로 이름을 설정하였는데 parentId가 더 나은가요?

ReadAlien replied:
json 데이터로 넘어오는 값이 parent 객체가 아니라 parentId로 받고 있어서 ReplyCreate request에선 parentId가 더 적합한 거 같아요

ReadAlien commented on src/main/java/com/gagoo/thiscoding/domain/maria/reply/infrastructure/ReplyEntity.java:
대댓글 기능으로 parent 컬럼이 필요한 거로 보이는데 EAGER 로딩은 페이징 처리 시 모든 연관 데이터를 한번에 가져와 성능 이슈가 발생할 수 있어서 LAZY 로딩으로 변경해서 실제 필요한 시점에만 조회되게 바꾸는 게 좋을 거 같아요

cyj1127 replied:
LAZY로 변경 후 무한 재귀 현상을 JsonIgnore 어노테이션을 통해 해결하였습니다.

ReadAlien commented on src/main/java/com/gagoo/thiscoding/domain/maria/reply/service/ReplyServiceImpl.java:
null체크는 Optional을 사용해서 안전하게 처리하고 데이터 변환 과정은 map 메서드 사용해서 메서드체이닝 방식으로 가독성 향상시키면 좋을 거 같아요

cyj1127 replied:
말씀하신대로 불안한 삼항연산자보단 Optional.ofNullable을 통해 댓글과 대댓글의 경우에 따라 분기하여 null허용을 해서 예외처리 진행하였고, 데이터 변환 과정은 순수도메인 유지하면서 필요없는 부분이라 삭제하였습니다.

ReadAlien reviewed (commented):
리뷰남겼습니다 확인해주세요

Commit:
refactor:Manager QueryDsl, Domain 변경

순수도메인 유지를 위해 ReplyEntity > Reply타입으로 변경

ReplyEntity parent FetchType을 EAGER > LAZY로 변경 및 JsonIgnore를 통해 무한 재귀 현상 해결

ReplyService에서 create메서드 삼항연산자 > Optional.ofNullable을 통해 댓글,대댓글 분기 및 안전하게 예외처리

QueryDsl 쿼리문에서 parent대신 parent.id로 불필요한 user정보 조회를 하지 않도록 수정

Commit:
feat:셀프 조인을 통한 Qna 대댓글 구현

Optional.ofNullable 메서드로 parentI가 존재하는 경우에만 findById 메서드 실행하여 오류 해결.

댓글 request는 Long타입으로 id를 받기에 parentId로 설정하고 그 외 엔티티 및 도메인은 부모댓글 객체이기에 parent로 변수명 설정

nameisris reviewed (approved):
필드명 변경 및 null 값 처리 등의 내용 확인했습니다.

nameisris commented on src/main/java/com/gagoo/thiscoding/domain/maria/reply/service/dto/ReplyList.java:
```diff
@@ -2,22 +2,27 @@
+    public ReplyList(Long replyId, Long parent, String nickname, String profileImage, String content, LocalDateTime createDate) {
         this.replyId = replyId;
-        this.parentId = parentId;
+        this.parentId = parent;
```
Long 타입으로 parentId라고 선언된 변수에 대해, Builder에서 넘겨주는 파라미터 명을 parent로 명시한 이유가 있나요/?

cyj1127 replied:
QueryDsl로 DB에 접근시엔 parent객체를 뽑아오고 다시 ReplyList를 통해 response로 전달할 땐 parent.id로 부모객체의 id값을 넘겨주기에 parentId로 설정하였습니다.

cyj1127 reviewed (commented):
리뷰 남겼습니다.
